cmd/storage-class.go

```diff
@@ -0,0 +1,252 @@
+		return errors.New("Standard storage class parity disks should be less than or equal to " + strconv.Itoa(len(globalEndpoints)/2))
+	}
+
+	return nil
+}
+
+// Returns the data and parity drive count based on storage class
+// If storage class is set using the env vars MINIO_STORAGE_CLASS_RRS and MINIO_STORAGE_CLASS_STANDARD
+// -- corresponding values are returned
+// If storage class is not set using environment variables, default values are returned
+// -- Default for Reduced Redundancy Storage class is, parity = 2 and data = N-Parity
+// -- Default for Standard Storage class is, parity = N/2, data = N/2
+// If storage class is not present in metadata, default value is data = N/2, parity = N/2
+func getDrivesCount(sc string, disks []StorageAPI) (data, parity int) {
+	totalDisks := len(disks)
+	switch sc {
+	case reducedRedundancyStorageClass:
+		if globalRRStorageClass.Parity != 0 {
+			// set the rrs parity if available
+			parity = globalRRStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = defaultRRSParity
+		}
+	case standardStorageClass:
+		if globalStandardStorageClass.Parity != 0 {
+			// set the standard parity if available
+			parity = globalStandardStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = totalDisks / 2
+		}
+	default:
+		parity = totalDisks / 2
+	}
+	// data is always totalDisks - parity
+	data = totalDisks - parity
+	return data, parity
+}
+
+func commonBlocks(partsMetaData []xlMetaV1, diskType string) (blocks, count int) {
+	diskCountMap := make(map[int]int)
+
+	switch diskType {
+	case dataBlocks:
+		for _, xlMeta := range partsMetaData {
+			if xlMeta.Erasure.DataBlocks == 0 {
+				continue
+			}
+			diskCountMap[xlMeta.Erasure.DataBlocks]++
+		}
+	case parityBlocks:
+		for _, xlMeta := range partsMetaData {
+			if xlMeta.Erasure.ParityBlocks == 0 {
+				continue
+			}
+			diskCountMap[xlMeta.Erasure.ParityBlocks]++
+		}
+	}
+
+	max := 0
+	for dataBlocks, count := range diskCountMap {
+		if max < count {
+			max = count
+			blocks = dataBlocks
+		}
+	}
+	return blocks, max
+}
+
+// commonDataBlocks returns the maximally occurring dataBlocks and its count from xlMetaV1 slice.
+func commonDataBlocks(partsMetaData []xlMetaV1) (dataBlock, count int) {
+	return commonBlocks(partsMetaData, dataBlocks)
+}
+
+// commonParityBlocks returns the maximally occurring parityBlocks and its count from xlMetaV1 slice.
+func commonParityBlocks(partsMetaData []xlMetaV1) (parityBlock, count int) {
+	return commonBlocks(partsMetaData, parityBlocks)
+}
+
+// Returns per object readQuorum and writeQuorum
+// readQuorum is the minimum required disks to read data.
+// writeQuorum is the minimum required disks to write data.
+func objectQuorumFromMeta(xl xlObjects, partsMetaData []xlMetaV1, errs []error) (objectReadQuorum, objectWriteQuorum int, err error) {
+
+	// get the latest update XL Metadata
+	latestXLMeta := listLatestXLMeta(partsMetaData, errs)
+
```

Comment:
this can be simplified as,

```go
emptyMeta := xlMetaV1{}
count := 0
for _, xlMeta := range latestXLMeta {
  if xlMeta == emptyMeta {
    continue
  }
  count ++
  if validXLMeta == emptyMeta  {
    validXLMeta = xlMeta
  }
}
if validXLMeta == emptyMeta || count < validXLMeta.Erasure.DataBlocks {
  return 0, 0, errXLReadQuorum
}
return validXLMeta.Erasure.DataBlocks, validXLMeta.Erasure.DataBlocks+1, nil
```